src/hotspot/os/linux/os_linux.cpp

```diff
@@ -4701,6 +4711,15 @@ jint os::init_2(void) {
     FLAG_SET_DEFAULT(UseCodeCacheFlushing, false);
   }
 
+  // Override the timer slack value if needed. The adjustment for the main
+  // thread will establish the setting for child threads, which would be
+  // most threads in JDK/JVM.
+  if (TimerSlack > 0) {
+    if (prctl(PR_SET_TIMERSLACK, TimerSlack) < 0) {
+      vm_exit_during_initialization("Setting timer slack failed: %s", os::strerror(errno));
+    }
+  }
+
```

Comment:
Is this enough? The API reads as if this should happen for every thread individually. 

Or do threads inherit this override from their parent threads? So you only set it once for main thread? 

But even then, you'd need to do it at least for every thread that gets attached, since their parentage is unknown.

This raises another question, if a thread is detached, should we restore the original timer slack?

Reply:
> Or do threads inherit this override from their parent threads? So you only set it once for main thread?

Exactly. And my understanding is that we are only concerned with the main set of Java threads not any natively attached threads. If you have native threads then they can set their own values. We should not be second-guessing or trying to save/restore.

Reply:
> > Or do threads inherit this override from their parent threads? So you only set it once for main thread?
> 
> Exactly. And my understanding is that we are only concerned with the main set of Java threads not any natively attached threads. If you have native threads then they can set their own values. We should not be second-guessing or trying to save/restore.

That seems an strange omission and at odds with other runtime decisions we force upon attached threads, e.g. stack size / guard pages.

If that decision is deliberate, we should at least document it in the flag description.

Reply:
??? We can't affect either the stacksize, nor guard pages of an existing native thread!

Reply:
> ??? We can't affect either the stacksize, nor guard pages of an existing native thread!

No, but we place guard pages into the stacks of attached threads. We also do other things, e.g. set the signal mask, or set the FPU control word.

My point is that we attempt to minimize the differences between attached threads and created threads. Which makes sense. Therefore it feels odd that Java APIs that depend on blocking syscalls should show different behavior depending on whether they are called in a created or in an attached thread. The developer of the Java code may not even know. You may run with hotspot embedded into some custom launcher, and the launcher doing its own thread pooling.

Reply:
We only do to attached threads what we need to do to enable them to run Java - and those things should not affect them once they have detached. We don't mess with attached thread names for example. You can make the case either way for this - you may want the attached thread to behave exactly like a native Java thread, or you may be upset that the behaviour of an attached threads changes when executing non Java code because of some action attaching took. So I don't have a concern that native threads are not affected by this experimental flag at least at this initial stage of experimentation.

Reply:
You are saying we do the minimum needed to run java, but leave them alone otherwise? Okay, I can see the logic in that.